### DISCUSSION TITLE
Is QUIT expected to THROW?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#116] [Q&A]
The Forth 200x Draft 19.1 section 6.1.2050 QUIT makes no mention of THROWing a code, while Table 9.1: THROW code assignments assigns a code for QUIT that implies that it might do so, but 9.6.2 Exception Extension words has not extension for QUIT to -56 THROW.

Is this an oversight?

**Comment by ruv** [comment]
Yes, [Table 9.1](https://forth-standard.org/standard/exception#table:throw) is inconsistent, and it's a [known problem](https://groups.google.com/g/comp.lang.forth/c/4pRd4VBPfJ8/m/cKbsoclsBgAJ), which should be fixed in the next version of the Standard.

In many rows of this table a throw code is reserved, but the corresponding ambiguous condition is not specified.  For example, for code `-56` — there is no such an ambiguous condition as "QUIT", and the glossary entry for the word [`QUIT`](https://forth-standard.org/standard/core/QUIT) doesn't mention any ambiguous condition, and the Exception word set doesn't update the glossary entry for `QUIT` to add an ambiguous condition. This table was populated by new throw codes that are ior codes from the proposal "[Throw IORs](http://forth200x.org/throw-iors.html)" ([copy](https://rawcdn.githack.com/Forth-Standard/forth200x/59729ede5e277bd08a5ba7721e22b61c7fc2a99c/throw-iors.html)), [accepted in 2007](http://forth200x.org/meetings/2007.html).


On the other hand, throw codes from Table 9.1 should be only used if a system **choosing** to execute `THROW` in a corresponding ambiguous condition. But the system **is allowed** to not execute `THROW` in the case of an ambiguous condition, and may take [another action](https://forth-standard.org/standard/usage#usage:ambiguous).

So it's not expected that `QUIT` _should_ throw an exception in the case of an ambiguous condition.

----

Concerning the cases when an exception is thrown and behavior is well defined.

Usually, `QUIT` is implemented in such a way that it can be only interrupted by the end of the input stream, or via [`BYE`](https://forth-standard.org/standard/tools/BYE). But apparently it doesn't meet the current wording of the specification for `QUIT` and `THROW`, since `QUIT` should not prevent propagation of exceptions.

**Reply by ruv** [reply]
Mitch Bradley, who originated Table 9.1, [wrote](https://groups.google.com/g/comp.lang.forth/c/P-naCsYjmGo/m/w-8fY3pE76kJ) in 1992 in comp.lang.forth:

>  the description of `THROW` says that, in the absence of explicitly-installed `CATCH` handlers, the behavior is like `ABORT` which eventually does the function of `QUIT`. An implementation of `QUIT` could do a "preemptive CATCH" and still be within the standard. This isn't surprising; a program that calls `QUIT` cannot expect to ever regain control anyway.

I.e., that `QUIT` under the hood may catch and don't re-throw the exception if there is no a user-defined exception frame at the moment.

But there is nothing concerning the throw code `-56`

**Reply by ruv** [reply]
@MitchBradley, could you please clarify how the throw code `-56` can be used?

**Comment by MitchBradley** [comment]
```
: top-level-app-word  ( -- )
   ['] do-app catch ?dup -56 = if  ." Exited via QUIT; stopping program" cr bye else ." Exited some other way" cr then
;
```
Maybe it is not a standard program, but it could potentially be useful.

History:
Keep in mind that we did not have infinite time to complete the standard.  CATCH and THROW were added relatively late in the multi-year deliberations, about the time when everybody was exhausted with the process.  The need for exception handling had been identified very early, but every early proposal had serious problems, with syntactic and control flow implications that were very un-Forth-like and thus interfered with nearly every other aspect of the standard.

At a meeting at Johns Hopkins, I was lying in the bathtub one evening and the basic structure of CATCH and THROW came to me.  I wrote down a basic implementation that night and showed it to Dean Sanderson the next day.  Dean liked it, and if you know Dean, that is high praise.  So we finally had a way forward for the lingering exception problem. The standard was nearly done so we did not have a lot of time to nitpick every detail.  Also, since the idea was so new, we did not want to risk integrating it deeply into the core, which is why it is in an extension wordset.

I quickly added it to my own Forth systems to see what would happen.  The results were very encouraging.  It cleaned up a lot of tricky circular dependencies in the core code.  I also realized that it helped a lot with handling error conditions with the file access wordset.

So we went with it, and tried to integrate it into the text of the standard as well as we could, given the constraint that we needed to finish soon before our patience, and the patience of our companies that were funding trips to meeting venues, ran out.  We missed some things.  But, all things considered, I am proud of the result.  It has proven to be workable and useful.

**Reply by ruv** [reply]
Mitch, thank you for your story, very interesting!  From those time no one came up with anything better than catch-throw.

-----
Concerning `quit` and `-56`.
```
: quit -56 throw ;
```
is similar to 
```
: abort -1 throw ;
```
It would work, if `quit` have to empty the data stack.  But `quit` should not empty the data stack, so it cannot use `throw` that does it.
So, yes, it would be non standard.

**Reply by SirWumpus** [reply]
If `QUIT` throws -56 (similar argument for `ABORT`), what is the desired out come?  From what I've been working with I see:

* top level no catch frame, `THROW` lets `QUIT` behave  per its Core definition (empty return stack, set standard input, resume REPL);

* one or more catch frames, `THROW` ignores/ overrides Core `QUIT`, restore data/return stack depths from the catch frame, and pass -56 on data stack.

Are there alternative behaviours worth considering?

**Reply by ruv** [reply]
> what is the desired out come? 

In general, your description is correct for the case when no system's exception frames on the exception stack.  But if such exception frame exists, the behavior will be incorrect — since the data stack will be changed.

For example, if `included` is implemented in the following way:
```
: include-file ( i*x fileid -- j*x )
  dup >r ['] (include-file) catch r> close-file swap throw throw
;
: included ( i*x sd-filename -- j*x )
  r/o open-file throw include-file
;
```

And an included file contains `123 quit` in interpretation state, then this `123` will not be available on the data stack on REPL start.

Actually, even if user's code catches and re-throws the exception `-56`, the data stack items will be lost in REPL.

-----

`quit` that throws `-56` implies an implementation of `throw` in the following way:
```
: reset-input ( -- ) 0 set-source-id postpone [ ;

: throw ( ior -- )
    ?dup 0= if exit then
    handler @ if (throw) then \ jump to catch
    \ an exception frame is absent
    r0 set-rp \ empty the return stack
    case
        -56 of endof \ just quit
        >r s0 set-sp r> \ empty the data stack
        -1 of endof \ no error message
        -2 of error-message type cr endof \ show the provided message
        ." Error # " . cr \ show a common error message
    endcase
    reset-input begin refill while interpret repeat
;

: quit ( i*x -- ) -56 throw ;
: abort ( i*x -- ) -1 throw ;
: sthrow ( i*x sd.message -- ) set-error-message -2 throw ;
: abort" postpone s" postpone sthrow ; immediate
```

NB: it's an **incorrect** implementation for `quit` (update: regardless whether the system uses `catch` when it executes user's code or not — as shown [below](https://github.com/ForthHub/discussion/discussions/116#discussioncomment-3541822)).

**Reply by ruv** [reply]
It's possible to implement a system so it doesn't use `catch` by itself, but far easier to use `catch`. Then the actions that take place when an exception frame is formally absent can be dedicated to the bottom-most exception frame handler under the hood.

```
: throw ( ior -- )
    ?dup 0= if exit then
    handler @ if (throw) then \ jump to catch
    \ an exception frame is absent
    treat-fatal-error
;
: print-error-message ( ior -- )
    case
        0 of endof \ no error
        -1 of endof \ no error message
        -2 of error-message type cr endof \ show the provided message
        ." Error # " . cr \ show a common error message
    endcase
;
: reset-input ( -- ) 0 set-source-id postpone [ ;
: interpret-input ( i*x -- j*x )
  begin refill while interpret repeat
;
: main-repl \ the Forth system REPL entry point
    begin ['] interpret-input catch dup while reset-input print-error-message repeat bye
;
: quit ( i*x -- ) r0 set-rp handler 0! reset-input main-repl ;

```

**Reply by SirWumpus** [reply]
> what is the desired out come?

From my understanding of `CATCH` and `THROW` you would still want (most of) the actions of the `xt` to be preformed, catching any deep exceptions.  `['] QUIT CATCH` I would expect it to perform the core behaviour and then `-56 THROW` as the last thing, but this can't work because of the stack behaviours.

```
: quit ( i*x -- ) -56 throw ;
: abort ( i*x -- ) -1 throw ;
```
In both the core words are replaced by a simplistic definition that simply throws a code and does nothing else.  Yet 9.6.2.0670 ABORT (similar for `ABORT"`) say

> Extend the semantics of 6.1.0670 ABORT to be ...

And I would presume for some future addition of a `QUIT` (exception ext) would also extend / append to the core definition, not replace it.  This is where all the confusion I think lies.  

At one time I had wanted a means to achieve emptying data and/or return stacks, reset input source to standard input, but not resume the REPL, and there does not appear to a portable way to do this. I ended up defining implementation specific words using implementation specific primitives.

I think there are clearly some holes in the draft:

* Catching `QUIT` and `ABORT` - what is expected?  Maybe best to punt and claim "undefined behaviour".
* Maybe new standard words that break down `QUIT` and `ABORT` into smaller fine grain parts, eg. DROPALL, RESET-INPUT, MAIN-REPL, that would avoid some of the issues in trying to catch `QUIT` and `ABORT` for their side-effects by letting the developer pick the needed parts.   I had considered an `RDROPALL` to empty the return stack, but that just duplicates part of the problem of catching `QUIT`.

**Reply by ruv** [reply]
> Yet 9.6.2.0670 ABORT (similar for `ABORT"`) say
 >>  Extend the semantics of 6.1.0670 ABORT to be ...

This wording is not quite well, I think. Since the semantics are actually replaced (not extended), and _formally_ recursive.
Probably, 9.6.2.0670 should not mention 6.1.0670 at all.   This will be fixed in the frame of [making the Exception word set mandatory](https://github.com/ForthHub/standard-evolution/issues/1).

**Reply by ruv** [reply]
> Catching `QUIT` and `ABORT` - what is expected? Maybe best to punt and claim "undefined behaviour".

Originally, in the bare metal Forth systems, Forth's REPL played role of a shell/terminal, and the idea of `QUIT` is that it **returns control** from a program to the terminal. It's like [`BYE`](https://forth-standard.org/standard/tools/BYE) in hosted Forth systems, [ExitProcess](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess) in Windows API, or [exit](https://pubs.opengroup.org/onlinepubs/9699919799/functions/exit.html) in POSIX. These functions don't return control to the caller regardless whether you wrap them with catch or not.

So `' quit catch` doesn't return control to the caller, as well as `' quit execute` or `' bye execute`.  And this behavior is well defined as far as some ambiguous condition is not met.

Catching `ABORT` just returns `-1`, i.e. `t{ ' abort catch -> -1 }`.  This behavior is well defined too.

**Reply by ruv** [reply]
>  fine grain parts, eg. `DROPALL`, `RESET-INPUT`, `MAIN-REPL`

By the way, `DROPALL` can be implemented as:

```
: dropall ( i*x -- ) depth 0 ?do drop loop ;
```

REPL can be also implemented in a standard way.  Also, a program can read from the user input device via [`ACCEPT`](https://forth-standard.org/standard/core/ACCEPT), but cannot forcefully  make the user input device the input source. The only way is to catch the point where the user input device is the input source and then return or throw to this point.

Concerning `RDROPALL` — it's difficult to specify semantics consistently for such a word.

**Reply by SirWumpus** [reply]
`RDROPALL` has the same problem as `QUIT` and `ABORT`.  I removed it from my implementation shortly after toying with it, cause I don't think there are any valid semantics for it.

**Comment by SirWumpus** [comment]
> So `' quit catch` doesn't return control to the caller, as well as `' quit execute` or `' bye execute`.

Sounds like something that needs some comment in the Rationale:

```
' QUIT CATCH
' QUIT EXECUTE

Never returns to the caller.  In some implementations, the throw code -56 in Table 9.1 may be
 used internally to transfer control to `THROW`, which in turn implements the `QUIT` semantics. 
```

**Reply by ruv** [reply]
>  the throw code -56 in Table 9.1 may be used internally to transfer control to `THROW`, which in turn implements the `QUIT` semantics

It seems impossibly to implement correctly, since when a user catches and re-throws an exception, the data stack depth can be decreased, and if a throw code is `-56`, then the arguments of `QUIT` will be lost.

**Reply by SirWumpus** [reply]
> ... then the arguments of `QUIT` will be lost.

But the draft does not indicate any arguments for `QUIT`.

**Reply by ruv** [reply]
> the draft does not indicate any arguments for `QUIT`.

All (and only) items on the data stack before `QUIT` is called shall be available in REPL. This feature is used for debugging purposes.

**Reply by SirWumpus** [reply]
Yes.  I have used it for such in the past, but it is not a requirement.

* A `QUIT` with no prior `CATCH`/`THROW` (other than possibly an internal -56 suggested above) would not disturb the data stack.
* If there are intermediate user (+ve codes) or exception `CATCH`/`THROW` followed by `QUIT`, then yes the stack depth will have changed along the way.  Any debugging / error handling of an intermediate `CATCH` would have to happen at that particular point before the next `CATCH`.  I don't see this a problem given the behaviour of `CATCH`/`THROW`.

**Reply by ruv** [reply]
>  I don't see this a problem given the behaviour of `CATCH`/`THROW`.

If some standard program shows a different behavior on a system that implements `QUIT` in a THROW-based way, then this implementation for `QUIT` is incorrect (not standard compliant).

In a standard compliant implementation **none of program's code is executed** after call of `QUIT`.  In a THROW-based implementation some program's code can be executed and show a different behavior. Hence a THROW-based implementation for `QUIT` is not standard.

**Reply by SirWumpus** [reply]
I would expect a THROW-based implementation would confine QUIT within itself.  I don't follow how / why a program's code could be executed during the QUIT-THROW-flow.

**Reply by ruv** [reply]
>  I don't follow how / why a program's code could be executed during the QUIT-THROW-flow.

I showed an implementation example for THROW-based `QUIT` in my [comment above](https://github.com/ForthHub/discussion/discussions/116#discussioncomment-3490340).

Let's consider the following program:
```
: foo [: 123 quit ;] catch 456 . throw ;  foo
```

- In a standard system this program doesn't print anything and we have `123` on the data stack in REPL.
- In a system having a THROW-based `QUIT` this program prints `456` and we have nothing on the data stack in REPL — since `THROW` (inside `QUIT`) drops stack items, and the fragment `456 .` is executed **after** the call to `QUIT` was made.

This program does not get into any ambiguous condition and doesn't detect any system's peculiarities, so all standard systems should show the same behavior by this program.

-----

On the other hand, the following program should print `-56 123`  on any standard system:
```
:noname 123 [: 456 -56 throw ;] catch . . ;  execute
```

**Reply by SirWumpus** [reply]
Thank you for that.  Took a moment, but appears to make sense now.
